elastic/beats
# [Metricbeat] Discard partial Logstash pipeline documents

## What does this PR do?

This PR closes https://github.com/elastic/kibana/issues/133230.

This PR is the Metricbeat module side of a three pronged fix to a race condition between Metricbeat and the Logstash APIs, which leads to partial data being ingested.

I've also updated the `710` fixture, as it had data from `7.8.1`.

It looks like these files haven't been linted before as there are a lot of linting errors and the linter is (AFAIK) pretty new. I haven't fixed these as they are heavily unrelated and belong to, mostly, only two main groups of issues.

## Why is it important?

This is important as this partial ingestion is not expected. There *should* always be a `hash` assigned (at least after version 7.3.0). This leads to the Stack Monitoring UI breaking as we can't rely on the data containing the properties needed.

## Checklist

- [x] My code follows the style guidelines of this project (I think so, couldn't find a STYLEGUIDE file)

## Author's Checklist

## How to test this PR locally

Note: I have also left a couple of inline code / review comments.

- Follow the docs to setup your environment to build Beats (here: https://www.elastic.co/guide/en/beats/devguide/current/beats-contributing.html#setting-up-dev-environment)

- Checkout this branch, and then`cd metricbeat`, and then `mage build`.

- Make a Metricbeat config file that contains the following (change the hosts and credentials as needed, this is an example). The `period: 1s` is important so we can capture the race condition:

```yml
http.enabled: true
metricbeat.modules:
  - module: system

  - module: logstash
    xpack.enabled: true
    period: 1s
    hosts: [ "localhost:9600" ]

output.elasticsearch:
  hosts: [ "localhost:9200" ]
  username: "elastic"
  password: "changeme"
```

- Then run `./metricbeat -e -c PATH_TO_YOUR_METRICBEAT_YML_FILE`

- Run Logstash. Docker works reliably in my testing with the `1s` period.

```
docker run --name logstash \                
--pull always --rm \
--hostname=logstash \
--publish=9600:9600 \
--volume="<PATH TO YOUR LOGSTASH YML>:/usr/share/logstash/config/logstash.yml:ro" \
docker.elastic.co/logstash/logstash:master-SNAPSHOT \
-e 'input { java_generator { eps => 1 } } output { file { path => "/dev/null" } }'
```

- The Beat should now poll at a time when a partial result is returned from the Logstash API. It should then discard this document, and it should not be ingested. You can verify this in the logs.

<img width="1791" alt="Screenshot 2022-06-12 at 00 51 09" src="https://user-images.githubusercontent.com/471693/173446137-6aa3defe-0bc8-41ac-8a7d-13bdbc6cdf2e.png">

(See the top two lines)

You can of course also verify this by querying Elasticsearch too for documents with `hash: ''`.


## Related issues

Closes https://github.com/elastic/kibana/issues/133230.

## Screenshots

An example of a partial result:

![Screenshot 2022-06-07 at 17 42 56](https://user-images.githubusercontent.com/471693/173446384-5fd46b03-66a7-4d2e-8f0f-da8b613f9682.png)

It's important to note that the Logstash API doesn't send back `hash: ''`, it doesn't send the key at all, but the unmarshalling of the JSON will assign an empty string as the property is defined on the struct.

After the unmarshal we see something like this:

![Screenshot 2022-06-07 at 18 02 24](https://user-images.githubusercontent.com/471693/173447227-d3b2ee75-2240-4971-94df-f27dfc32cc95.png)


## Logs

<img width="1791" alt="Screenshot 2022-06-12 at 00 51 09" src="https://user-images.githubusercontent.com/471693/173446280-c4d35459-2b13-418b-8005-7da0cc2bde50.png">
